storm-core/src/jvm/backtype/storm/scheduler/resource/User.java

```diff
@@ -0,0 +1,344 @@
+    }
+
+    public void addTopologyToInvalidQueue(TopologyDetails topo) {
+        this.invalidQueue.add(topo);
+    }
+
+    public Set<TopologyDetails> getTopologiesRunning() {
+        TreeSet<TopologyDetails> ret = new TreeSet<TopologyDetails>(new PQsortByPriorityAndSubmittionTime());
+        ret.addAll(this.runningQueue);
+        return ret;
+    }
+
+    public Set<TopologyDetails> getTopologiesAttempted() {
+        TreeSet<TopologyDetails> ret = new TreeSet<TopologyDetails>(new PQsortByPriorityAndSubmittionTime());
+        ret.addAll(this.attemptedQueue);
+        return ret;
+    }
+
+    public Set<TopologyDetails> getTopologiesInvalid() {
+        TreeSet<TopologyDetails> ret = new TreeSet<TopologyDetails>(new PQsortByPriorityAndSubmittionTime());
+        ret.addAll(this.invalidQueue);
+        return ret;
+    }
+
+    public Map<String, Number> getResourcePool() {
+        if (this.resourcePool != null) {
+            return new HashMap<String, Number>(this.resourcePool);
+        }
+        return null;
+    }
+
+    public void moveTopoFromPendingToRunning(TopologyDetails topo, Cluster cluster) {
+        moveTopology(topo, this.pendingQueue, "pending", this.runningQueue, "running");
+        if (cluster != null) {
+            cluster.setStatus(topo.getId(), "Fully Scheduled");
+        }
+    }
+
+    public void moveTopoFromPendingToRunning(TopologyDetails topo) {
+        this.moveTopoFromPendingToRunning(topo, null);
+    }
+
+
+    public void moveTopoFromPendingToAttempted(TopologyDetails topo, Cluster cluster) {
+        moveTopology(topo, this.pendingQueue, "pending", this.attemptedQueue, "attempted");
+        if (cluster != null) {
+            cluster.setStatus(topo.getId(), "Scheduling Attempted but Failed");
+        }
+    }
+
+    public void moveTopoFromPendingToAttempted(TopologyDetails topo) {
+        this.moveTopoFromPendingToAttempted(topo, null);
+    }
+
+
+    public void moveTopoFromPendingToInvalid(TopologyDetails topo, Cluster cluster) {
+        moveTopology(topo, this.pendingQueue, "pending", this.invalidQueue, "invalid");
+        if (cluster != null) {
+            cluster.setStatus(topo.getId(), "Scheduling Attempted but topology is invalid");
+        }
+    }
+
+    public void moveTopoFromPendingToInvalid(TopologyDetails topo) {
+        this.moveTopoFromPendingToInvalid(topo, null);
+    }
+
+
+    public void moveTopoFromRunningToPending(TopologyDetails topo, Cluster cluster) {
+        moveTopology(topo, this.runningQueue, "running", this.pendingQueue, "pending");
+        if (cluster != null) {
+            cluster.setStatus(topo.getId(), "Scheduling Pending");
+        }
+    }
+
+    public void moveTopoFromRunningToPending(TopologyDetails topo) {
+        this.moveTopoFromRunningToPending(topo, null);
+    }
+
+
+    private void moveTopology(TopologyDetails topo, Set<TopologyDetails> src, String srcName, Set<TopologyDetails> dest, String destName) {
```

Comment:
I think if we exposed a cleaned up version of this method a lot of the error handling logic in the scheduler could be turned into a single method.